### PR TITLE
Support ERM Non-destructive push for User Roles

### PR DESCRIPTION
**Pull request by mjriley**
## Technical Summary

This is the first in a series of PRs meant to support ERM non-destructive push. Associated ticket can be found [here](https://dimagi-dev.atlassian.net/browse/SAAS-13622).  This PR contains generic updates as well as the bulk of the code to support this feature for User Roles.

The pain point this work is intended to solve is that downstream domains sometimes want to maintain independent data from upstream domains. If a user were to sync with the previous ERM workflow, the sync would always overwrite any data from the downstream domain with no warning. This has now been changed, so that:
1) previously synced data continues to sync as normal
2) conflicting/matching data raises an error
3) users have the ability to overwrite that conflict anyway (will be included in a future PR).

I'm started here, so that the PRs are easier to review and digest. There are future PRs for:
- Custom Data Fields
- Auto Update Rules
- Fixtures
- Force Push
- Keywords
- Misc

### Safety story
This has been locally tested and has been on staging for weeks or months.

### Automated test coverage

New test suites have been created to cover existing functionality, and then augmented to handle the new ERM requirements.

### QA Plan

https://dimagi-dev.atlassian.net/browse/QA-5375

### Migrations

- [x] The migrations in this code can be safely applied first independently of the code

I believe this to be true. `upstream_id` fields have been added where necessary, and I believe the extra field is harmless.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/linked_domain/tasks.py [hunk 1]
```diff
@@ -272,6 +272,10 @@ def release_domain(upstream_domain, downstream_domain, username, models, build_a
 def send_linked_domain_release_email(results, upstream_domain, username, models, downstream_domains):
     manager = ReleaseManager(upstream_domain, username)
 
+    # # chord sends a list of results only if there were multiple tasks
```
Comment (biyeun) [inline corehq/apps/linked_domain/tasks.py:275]: it looks like this commit just added commented out code. was that intentional?

File: corehq/apps/users/models.py [hunk 2]
```diff
@@ -558,6 +558,18 @@ def is_domain_admin(self, domain=None):
         else:
             return False
 
+    # I'm not sure this is the correct place for this, as it turns a generic module
+    #  into one that knows about ERM specifics. It might make more sense to move this into
+    #  into the domain membership module, as that module knows about specific permissions.
+    # However, because this class is the barrier between the user and domain membership,
+    # exposing new functionality on domain membership wouldn't change the problem.
```
Comment (millerdev) [inline corehq/apps/users/models.py:565]: Could it be implemented as a stand-alone function in the domain membership module?
```py
def can_edit_linked_data(user, domain):
    ...
```

File: corehq/apps/users/templates/users/partials/linked_roles_modal.html [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+            <span aria-hidden="true">&times;</span>
+            </button>
+            <h3 class="modal-title">
+                <i class="fa fa-link"></i>
+                Unlock Linked Roles?
+            </h3>
+        </div>
+        <div class="modal-body">
+            <p>Linked Roles are roles that are controlled by a separate project space linked to this project.
+                Click here to learn more.</p>
```
Comment (millerdev) [inline corehq/apps/users/templates/users/partials/linked_roles_modal.html:18]: Should "Click here" be a link?

Reading on below, I noticed that the convention used in what look like similar "learn more" links is
```html
<a href='...' target='_blank'>Learn more</a>.
```
Would it make sense to have this one be consistent with other similar links?

File: corehq/apps/linked_domain/updates.py [hunk 4]
```diff
@@ -299,15 +307,47 @@ def update_user_roles(domain_link):
...
+        return False
+
+    default_permissions = UserRolePresets.INITIAL_ROLES[downstream_role.name]()
+    upstream_permissions = HqPermissions.wrap(upstream_role_json['permissions'])
+    if (upstream_permissions == downstream_role.permissions == default_permissions):
+        return True
 
-def _get_matching_downstream_entry(upstream_role_json, downstream_roles):
-    return next((role for role in downstream_roles if role.upstream_id == upstream_role_json['_id']), None)
+    return False
```
Comment (millerdev) [inline corehq/apps/linked_domain/updates.py]: nit
```suggestion
    return upstream_permissions == downstream_role.permissions == default_permissions
```

**Review by millerdev — APPROVED** [review]
None of the comments are blocking.

File: corehq/apps/users/templates/users/partials/linked_roles_modal.html [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+            <p>Changes made to Linked Roles may be overwritten by the next updates from the linked project space,
+                and may also affect solutions on this project space.</p>
```
Comment (gherceg) [inline corehq/apps/users/templates/users/partials/linked_roles_modal.html:20]: I find the wording "may also affect solutions on this project space" a bit strange. What is meant by "solutions" here?

File: corehq/apps/linked_domain/updates.py [hunk 6]
```diff
@@ -259,37 +261,43 @@ def update_user_roles(domain_link):
...
 
     is_embedded_tableau_enabled = EMBEDDED_TABLEAU.enabled(domain_link.linked_domain)
     if is_embedded_tableau_enabled:
         visualizations_for_linked_domain = TableauVisualization.objects.filter(
             domain=domain_link.linked_domain)
+    failed_updates = []
+    successful_updates = []
 
     # Update downstream roles based on upstream roles
     for upstream_role_def in master_results:
```
Comment (gherceg) [inline corehq/apps/linked_domain/updates.py:277]: nit: this should probably be changed to `upstream_roles` or something like that

File: corehq/apps/linked_domain/updates.py [hunk 7]
```diff
@@ -299,15 +307,47 @@ def update_user_roles(domain_link):
...
+    for downstream_role in downstream_roles:
+        if upstream_role_json['_id'] == downstream_role.upstream_id:
+            matching_role = downstream_role
+        elif upstream_role_json['name'] == downstream_role.name:
+            if _is_default_built_in_role(upstream_role_json, downstream_role):
+                matching_role = downstream_role
+            else:
+                conflicting_role = downstream_role
+
+    return (matching_role, conflicting_role)
```
Comment (gherceg) [inline corehq/apps/linked_domain/updates.py:330]: nit: just curious why you chose this instead of a tuple that returned `(role, is_conflict)`

**Review by gherceg — APPROVED** [review]
⭐ I could read your tests for days. Looks great! Nothing blocking on my end.

File: corehq/apps/users/tests/test_models.py [hunk 8]
```diff
@@ -278,3 +278,17 @@ def test_normalize_edit_apps_implies_view_apps(self):
...
         self.assertTrue(permissions.view_apps)
+
+    def test_normalize_access_release_management_preserves_previous_edit_linked_config_value(self):
+        permissions = HqPermissions(access_release_management=True, edit_linked_configurations=True)
+        old_permissions = HqPermissions(edit_linked_configurations=False)
+        permissions.normalize(previous=old_permissions)
+
+        self.assertFalse(permissions.edit_linked_configurations)
+
+    def test_normalize_disabled_release_management_uses_edit_linked_config_value(self):
```
[the remark this answers is not in the document]
Comment (mjriley) [reply corehq/apps/users/tests/test_models.py:289]: I think so. I believe the Release Management permission is meant to be like the ERM superuser, implicitly including any granular permissions. The Edit Linked Config permissions is intended to be one of those granular permissions when you don't want the user to have the full permissions of Release Management..

File: corehq/apps/linked_domain/updates.py (hunk 7 above)
Comment (mjriley) [reply corehq/apps/linked_domain/updates.py:330]: Good catch. I believe I initially implemented it as you're suggesting, then ended up needing the specific name of the role that conflicted. I don't think that's necessary anymore, so I'm going to adopt what you're suggesting here

File: corehq/apps/users/tests/test_models.py [hunk 9]
```diff
@@ -209,3 +210,71 @@ def test_uses_max_from_heartbeat(self):
...
+        permissions = HqPermissions(edit_web_users=True, view_web_users=False)
+        permissions.normalize()
+
+        self.assertTrue(permissions.view_web_users)
+
+    def test_normalize_edit_commcare_users_implies_view_users(self):
+        permissions = HqPermissions(edit_commcare_users=True, view_commcare_users=False)
+        permissions.normalize()
+
+        self.assertTrue(permissions.edit_commcare_users)
```
[the remark this answers is not in the document]
Comment (mjriley) [reply corehq/apps/users/tests/test_models.py:250]: Thanks, definitely an oversight from me

**Comment by mjriley** [comment]
As mentioned in the PR's description, this was 1 of 5 PRs to support ERM. Each of those PRs was individually approved before being merged into this one.

**Pull request merged**